frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -0,0 +1,708 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! The signed phase implementation.
+
+use crate::{
+	CompactOf, Config, ElectionCompute, Pallet, RawSolution, ReadySolution, SolutionOrSnapshotSize,
+	Weight, WeightInfo, QueuedSolution, SignedSubmissions,
+};
+use codec::{Encode, Decode, HasCompact};
+use frame_support::traits::{Currency, Get, OnUnbalanced, ReservableCurrency};
+use sp_arithmetic::traits::SaturatedConversion;
+use sp_npos_elections::{is_score_better, CompactSolution};
+use sp_runtime::{Perbill, RuntimeDebug, traits::Zero};
+use sp_std::{cmp::Ordering, vec::Vec};
+
+/// A raw, unchecked signed submission.
+///
+/// This is just a wrapper around [`RawSolution`] and some additional info.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, Default)]
+pub struct SignedSubmission<AccountId, Balance: HasCompact, CompactSolution> {
+	/// Who submitted this solution.
+	pub who: AccountId,
+	/// The deposit reserved for storing this solution.
+	pub deposit: Balance,
+	/// The reward that should be given to this solution, if chosen the as the final one.
+	pub reward: Balance,
+	/// The raw solution itself.
+	pub solution: RawSolution<CompactSolution>,
+}
+
+pub(crate) type BalanceOf<T> =
+	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+pub(crate) type PositiveImbalanceOf<T> = <<T as Config>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::PositiveImbalance;
+pub(crate) type NegativeImbalanceOf<T> = <<T as Config>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::NegativeImbalance;
+
+impl<T: Config> Pallet<T> {
+	/// Finish the signed phase. Process the signed submissions from best to worse until a valid one
+	/// is found, rewarding the best one and slashing the invalid ones along the way.
+	///
+	/// Returns true if we have a good solution in the signed phase.
+	///
+	/// This drains the [`SignedSubmissions`], potentially storing the best valid one in
+	/// [`QueuedSolution`].
+	pub fn finalize_signed_phase() -> (bool, Weight) {
+		let mut all_submissions: Vec<SignedSubmission<_, _, _>> = <SignedSubmissions<T>>::take();
+		let mut found_solution = false;
+		let mut weight = T::DbWeight::get().reads(1);
+
+		// Reverse the ordering of submissions: previously it was ordered such that high-scoring
+		// solutions have low indices. Now, the code flows more cleanly if high-scoring solutions
+		// have high indices.
+		all_submissions.reverse();
+
+		while let Some(best) = all_submissions.pop() {
+			let SignedSubmission { solution, who, deposit, reward } = best;
+			let active_voters = solution.compact.voter_count() as u32;
+			let feasibility_weight = {
```

Comment:
can use `feasibility_weight_of `

Reply:
No, unfortunately; that uses `solution.compact.unique_targets()`, while here we use `Self::desired_targets().unwrap_or_default()`.